Refactor the profiler to replace the timing prints

almaeder opened the pull request:
This PR refactors the old profiler:

- Used the assumptions of having only a single GPU to simplify the code
- Changed the profiling/profiling_range to have the same behavior as the prints (sync/time/sync/time/barrier/time)
- Replaces the prints with the profiler
- Dump the profiling results at the end of the CLI run
- Write the prints in the profiling to a separate file to clean up the outputs
- Simplify the resulting stats that are dumped

This PR should be merged only after GW-kpoints / better batching is done, since it will cause many conflicts.

almaeder committed:
Discontinue decorate_methods of the profiler

The decorate_methods is discontinued since the profiler
will only be used with manual labels to achieve profiling
at specific points. This also results in "utils.py" not beeing needed anymore.
Also, get_cuda_devices is moved since it is only used in "profiler.py".

almaeder committed:
Changed how profiler collects data

If cupy available, always get GPU times (no host times).
Other simplifying assumption: Only one device is used.
This should always be the case and not be changed in any near
future.

- The depth of the profiling is included to allow better printing
and flow diagram.
- Default to barrier afterwards to measure different between ranks.

almaeder committed:
Fixed bug: no comm input, correct sync

Comm input for the profiling range.
Comm syncronize first event to make sure that kernel before finished.

almaeder committed:
Added prints and use of comm_world

comm was aliasing with input variables
prints are useful and will be later piped to a
separate file.

almaeder committed:
Fixed the profile ranges and remove all .profile

The .profile are removed since there are placed everywhere.
Later they will only used sparsely.
The profile ranges where already at the correct places for P/S.

almaeder committed:
Replace manual timers with the profiler

This replaces all barriers, sync, prints in quatrex
with the profiler. It makes it much cleaner.

almaeder committed:
Write the timing outputs to a separate file.

The timings are by default writen to qtx_times.out.
Modified quatrex and the profiler for this.
Also dump the final stats after scba in the cli.
Add the files to gitignore.
Remove old profiler paramters from the config and
added the new ones.

almaeder committed:
Fixed the profiler tests

The tests were wrong due to changes to the profiler.
dump_stats has no inputs anymore since the paths are set with
set_parameters.

almaeder committed:
Remove profiler from files where it is not used

- The profiler was called in many files where it wasnt used anymore.

almaeder committed:
Remove `_get_cuda_devices` as only a single device per rank is assumed

- Previously, the profiler had functionality to profiler multiple devices per rank
- Additionally, complete the `Attributes` docstring

almaeder committed:
Remove extra RNG in BEYN

- Due to wrong rebase the older module level RNG was existing again.

almaeder committed:
Remove the bandedge comm time prints

- This print was still in the old style.
- The measurement of this is handled on a higher level and we do not need this fine grained measurement.

almaeder committed:
Remove sparsity pattern time prints

- Prints were still in the older style.
- They got replaced by the `profiler_range`.

almaeder commented:
@vetschn @alexnick83 the timing prints are currently appended if the output file already exists.
Are you okay with this behaviour, or should every `quatrex run` overwrite the timing print file?

vetschn commented on examples/w90/carbon-nanotube/gw-dist/quatrex_config.toml:
Did you change some parameters here? We really should stop tracking this with lfs...

almaeder replied:
`profiling_stats = false` was changed to `save_profiling_results = true`

almaeder replied:
We should exclude text files from LFS.

vetschn commented on src/qttools/greens_function_solver/rgf_dist.py:
```diff
@@ -270,179 +291,94 @@ def selected_solve(
+                xg_buffer_lower=xg_buffer_lower,
                 xg_buffer_upper=xg_buffer_upper,
                 xg_out=xg_out_,
-                selected_solve=True,
```
Diff is just cooked around here, or are there actual changes?

almaeder replied:
I believe the diff is cooked.

vetschn commented on src/qttools/profiling/profiler.py:
```diff
@@ -197,7 +222,17 @@ def __new__(cls):
-            cls._instance.devices = get_cuda_devices()
+            cls._instance.depth = -1
+            cls._instance.print_file = _OutputFile("qtx_times")
+            cls._instance.save_path = "qtx_times"
```
`quatrex_times` better? Everywhere else it is spelled out (like `quatrex_config.toml` etc.)

vetschn commented on src/qttools/profiling/profiler.py:
```diff
@@ -363,60 +349,82 @@ def profile(self, level: str = QTX_PROFILE_LEVEL):
-                    self._record_events(start_events)
+                    self.start_event.record(xp.cuda.get_current_stream())
+                    self.start_event.synchronize()
+                    self.start_event.record(xp.cuda.get_current_stream())
```
Yes, looks way better to assume only one device 👍

vetschn commented on src/qttools/profiling/profiler.py:
```diff
@@ -12,36 +12,44 @@
+    def write(self, message):
+        print(message, flush=True, file=self.file_handle)
+
+    def __delete__(self):
```
Are you sure you want `__delete__` here and not `__del__`? Could you quickly explain what this is really supposed to do?

almaeder replied:
Gemini :shaking-fist:. It should be `__del__`.

vetschn reviewed (changes requested):
Seems very nice! Only thing to me would be changing the name of the default output file to something like `quatrex_timings` or just `timings`. Also no file extension for this?

almaeder committed:
Rename profiling output files from 'qtx_times' to 'quatrex_times' for consistency

- Additionaly, pull out the extension print file extension.

almaeder committed:
Use profile range inside decorator

- Currently, the timing logic was duplicated, but it is better to maintain only thelogic inside profile_range.

vetschn reviewed (approved):
Looks great! I guess all that's left to do is make the linter happy

almaeder committed:
Strip the extension from the detected slurm output

- Else the print output file is `.out.*.out`

almaeder committed:
Combine both profiler paths to one

- Previously, there was a path for the print file and one for the saving the result.
- Both are now combined and the only difference is the extension.

almaeder committed:
Lint and format

- Previously, there were unused imports.
